pkg/controllermanager/controller/bastion/bastion_control.go

```diff
@@ -0,0 +1,193 @@
+// Copyright (c) 2021 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bastion
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	operationsv1alpha1 "github.com/gardener/gardener/pkg/apis/operations/v1alpha1"
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+func (c *Controller) bastionAdd(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.bastionQueue.Add(key)
+}
+
+func (c *Controller) bastionUpdate(_, newObj interface{}) {
+	c.bastionAdd(newObj)
+}
+
+func (c *Controller) bastionDelete(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		logger.Logger.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	c.bastionQueue.Add(key)
+}
+
+func (c *Controller) shootAdd(obj interface{}) {
+	shoot, ok := obj.(*gardencorev1beta1.Shoot)
+	if !ok {
+		return
+	}
+
+	// only shoot deletions should trigger this, so we can cleanup Bastions
+	if shoot.DeletionTimestamp == nil {
+		return
+	}
+
+	// list all bastions that reference this shoot
+	bastionList := operationsv1alpha1.BastionList{}
+	listOptions := client.ListOptions{
+		Namespace: shoot.Namespace,
+		// FieldSelector: fields.SelectorFromSet(fields.Set{operations.BastionShootName: shoot.Name}),
```

Comment:
@petersutter This should use the field selector, but I was not able to get it to work, hence why I went the slow road and filter in the loop below.

Reply:
I guess it's because only `BastionSeedName` is indexed
https://github.com/xrstf/gardener/blob/bastion-controller/pkg/registry/operations/bastion/storage/storage.go#L69

Reply:
I tried that, but since https://github.com/gardener/gardener/blob/master/vendor/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L313-L315 doesn't allow multiple TriggerFuncs, I thought I had to revert. Not sure if "find bastion by seed" or "find bastion by shoot" is more important.

Reply:
I would say that "find bastion by seed" is more important. Ok then leave it as is with the loop and just remove the fieldSelector comment

Reply:
Fixed.